lib/aliquot/payment.rb

```diff
@@ -0,0 +1,135 @@
+module Aliquot
+  ##
+  # A Payment represents a single payment using Google Pay.
+  # It is used to verify/decrypt the supplied token by using the shared secret,
+  # thus avoiding having knowledge of merchant primary keys.
+  class Payment
+    ##
+    # Parameters:
+    # token_string::  Google Pay token (JSON string)
+    # shared_secret:: Base64 encoded shared secret
+    # merchant_id::   Google Pay merchant ID ("merchant:<SOMETHING>")
+    # logger::        The logger to use. Default: Logger.new($stdout)
+    # signing_keys::  Formatted list of signing keys used to sign token contents.
+    #                 Otherwise a thread continuously updating google signing
+    #                 keys will be started.
+    def initialize(token_string, shared_secret, merchant_id, logger: Logger.new($stdout), signing_keys: nil)
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
+
+      validation = Aliquot::Validator::Token.new(JSON.parse(token_string))
+      validation.validate
+
+      @token = validation.output
+
+      @shared_secret = shared_secret
+      @merchant_id   = merchant_id
+      @signing_keys  = signing_keys
+    end
+
+    ##
+    # Validate and decrypt the token.
+    def process
+      unless valid_protocol_version?
+        raise Error, 'only ECv1 protocolVersion is supported'
+      end
+
+      unless valid_signature?
+        raise InvalidSignatureError
+      end
+
+      validator = Aliquot::Validator::SignedMessage.new(JSON.parse(@token['signedMessage']))
+      validator.validate
+      signed_message = validator.output
+
+      aes_key, mac_key = derive_keys(signed_message['ephemeralPublicKey'], @shared_secret, 'Google')
+
+      unless valid_mac?(mac_key, signed_message['encryptedMessage'], signed_message['tag'])
+        raise InvalidMacError
+      end
+
+      @message = JSON.parse(decrypt(aes_key, signed_message['encryptedMessage']))
```

Comment:
I think we should consider catching the potential json parse error and raise another (more descriptive) error. This should be tightly spec'ed IMO.